internal/cli/exec.go

```diff
@@ -16,6 +16,71 @@ import (
 
 type ExecCommand struct {
 	*baseCommand
+
+	flagInstanceId string
+}
+
+func (c *ExecCommand) targeted(args []string) int {
+	var exitCode int
+
+	client := c.project.Client()
+
+	err := c.DoApp(c.Ctx, func(ctx context.Context, app *clientpkg.App) error {
+		// We validate the users instance id first, since they're easy to mistype
+		// and we can give them quicker and better feedback by doing this up front.
+
+		resp, err := client.ListInstances(ctx, &pb.ListInstancesRequest{
+			Scope: &pb.ListInstancesRequest_Application_{
+				Application: &pb.ListInstancesRequest_Application{
+					Application: app.Ref(),
+				},
+			},
+		})
+		if err != nil {
+			app.UI.Output(clierrors.Humanize(err), terminal.WithErrorStyle())
+			return ErrSentinel
+		}
+
+		var found bool
+		for _, i := range resp.Instances {
+			if i.Id == c.flagInstanceId {
+				found = true
+			}
+		}
+
+		if !found {
+			app.UI.Output("Unable to find instance: %s", c.flagInstanceId, terminal.WithErrorStyle())
+			return ErrSentinel
+		}
+
+		// Ok, the instance id is fine, let's go ahead and have execclient do it's thing.
+
+		client := &execclient.Client{
+			Logger:     c.Log,
+			UI:         c.ui,
+			Context:    ctx,
+			Client:     client,
+			InstanceId: c.flagInstanceId,
+			Args:       args,
+			Stdin:      os.Stdin,
+			Stdout:     os.Stdout,
+			Stderr:     os.Stderr,
+		}
+
+		exitCode, err = client.Run()
+		if err != nil {
+			app.UI.Output(clierrors.Humanize(err), terminal.WithErrorStyle())
+			return ErrSentinel
+		}
+
+		return nil
+	})
```

Comment:
This is alright and all but I wonder if we should instead keep the same DoApp loop and just maybe add a couple functions we call out to to configure the `*execclient.Client`. I expect we'll have more targeting in the future (by deployment for example) and I don't want to set a precedent for just copying and pasting each time.

The actual instance lookup stuff looks fine!

Reply:
Yeah, I considered that. This change was the cleanest diff but I can move it inside DoApp.